### PR TITLE
Serialize store in PHP

### PR DESCRIPTION
**Pull request by DAreRodz**
## What

Adds store serialization and hydration.

## Why

The framework requires them. Developers should be able to send the initial state from the server as part of the SSR.

## How

- [x] Implementing a `WP_Directive_Store` class that contains the generated state.
- [x] Adding a `store( $data )` function to define the state on the block's render callback.
- [x] Adding a `evaluate( $path, $context )` function to access both the state and the context (like in JS)
- [x] Serializing and sending the store's content to the client in a `<script type="application/json">` tag.
- [x] Parsing the serialized store in the client and using it to initialize the store.
- [x] Rename `wpx` to `store`

### CODE REVIEW

File: phpunit/directives/utils/evaluate.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ */
+class Tests_Utils_Evaluate extends WP_UnitTestCase {
+	public function test_evaluate_function_should_access_state() {
+		// Init a simple store.
+		store(
+			array(
+				'state' => array(
+					'core' => array(
+						'a'      => 1,
+						'b'      => 2,
```
Comment (ockham) [inline phpunit/directives/utils/evaluate.php]: How about throwing in a different type value (to make sure `evaluate` gets the type right)? 😄 

```suggestion
						'b'      => true,
```

(and then `assertTrue` further down).

File: phpunit/directives/utils/evaluate.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		$this->assertSame( 3, evaluate( 'state.core.nested.c' ) );
+	}
+	public function test_evaluate_function_should_access_passed_context() {
+		$context = array(
+			'core' => array(
+				'a'      => 1,
+				'b'      => 2,
+				'nested' => array(
+					'c' => 3,
+				),
```
Comment (ockham) [inline phpunit/directives/utils/evaluate.php]: Maybe use a value for `$context` that's different from `$state`?

File: src/directives/attributes/wp-bind.php [hunk 3]
```diff
@@ -17,7 +17,7 @@ function process_wp_bind( $tags, $context ) {
 
 		// TODO: Properly parse $value.
```
Comment (ockham) [inline src/directives/attributes/wp-bind.php]: Let's maybe remove the `TODO` already -- it doesn't make sense to have it in every directive processor's individual code 😄 

```suggestion
```

File: src/directives/attributes/wp-class.php [hunk 4]
```diff
@@ -17,7 +17,7 @@ function process_wp_class( $tags, $context ) {
 
 		// TODO: Properly parse $value.
```
Comment (ockham) [inline src/directives/attributes/wp-class.php]: Same as https://github.com/WordPress/block-hydration-experiments/pull/147/files#r1101622012

```suggestion
```

File: src/directives/attributes/wp-style.php [hunk 5]
```diff
@@ -17,7 +17,7 @@ function process_wp_style( $tags, $context ) {
 
 		// TODO: Properly parse $value.
```
Comment (ockham) [inline src/directives/attributes/wp-style.php]: Same as https://github.com/WordPress/block-hydration-experiments/pull/147/files#r1101622012

```suggestion
```

File: src/directives/utils.php [hunk 6]
```diff
@@ -1,15 +1,26 @@
...
-		$result = $context;
-		foreach ( $path as $key ) {
-			$result = $result[ $key ];
+require_once  __DIR__ . '/class-wp-directive-store.php';
+
+function store( $data ) {
+	WP_Directive_Store::merge_data( $data );
+}
+
+function evaluate( string $path, array $context = array() ) {
```
Comment (ockham) [inline src/directives/utils.php:10]: Let's maybe add a TODO here.

```suggestion
// TODO: Implement evaluation of complex logical expressions.
function evaluate( string $path, array $context = array() ) {
```

File: wp-directives.php [hunk 7]
```diff
@@ -235,11 +238,24 @@ function wp_process_directives( $block_content ) {
...
+add_filter(
+	'render_block',
+	'wp_process_directives',
+	10,
+	1
+);
+
+// TODO: check if priority 9 is enough.
+// TODO: check if `wp_footer` is the most appropriate hook.
+add_action( 'wp_footer', WP_Directive_Store::render, 9 );
```
Comment (ockham) [inline wp-directives.php]: Ah, don't we normally use a different notation for invoking a class method? 🤔 

```suggestion
add_action( 'wp_footer', array( 'WP_Directive_Store', 'render' ), 9 );
```

File: e2e/store-tag.spec.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+		await page.goto('file://' + join(__dirname, 'store-tag-ok.html'));
+		const value = page.getByTestId('counter value');
+		const double = page.getByTestId('counter double');
+		const clicks = page.getByTestId('counter clicks');
+
+		await expect(value).toHaveText('3');
+		await expect(double).toHaveText('6');
+		await expect(clicks).toHaveText('0');
+
+		page.getByTestId('counter button').click();
```
Comment (ockham) [inline e2e/store-tag.spec.ts:15]: Can we also add unit test coverage for this functionality in the future (to test things in a more isolated way)? 😊 

I.e. for `wp-on:click`, we'd likely want to pass a [mock function](https://jestjs.io/docs/mock-function-api) and would check if it's really been called.

Or for `wp-bind`, we'd probably want to update a `state` value manually, and verify that the bound attribute is changed accordingly.

File: phpunit/directives/wp-directive-store.php [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+					'core' => array(
+						'a' => 'overwritten',
+					),
+				),
+			),
+			WP_Directive_Store::get_data()
+		);
+	}
+
+	public function test_store_should_be_correctly_render() {
```
Comment (ockham) [inline phpunit/directives/wp-directive-store.php]: Grammar nitpick 😊 

```suggestion
	public function test_store_should_be_correctly_rendered() {
```

**Review by ockham — APPROVED** [review]
I left a few minor notes, but LGTM in general! 🚢

File: wp-directives.php (hunk 7 above)
Comment (DAreRodz) [reply wp-directives.php]: I was sure there was a syntax for that, but I couldn't remember. 😄

File: e2e/store-tag.spec.ts (hunk 8 above)
Comment (DAreRodz) [reply e2e/store-tag.spec.ts:15]: Agree. We could do so in a future PR (also for other directives/functionalities). ☝️

**Comment by luisherranz** [comment]
Great work, David 🙂

---

For context here, we also need to merge the state that comes in the new page during client-side navigations. 

I've been talking with David and we think merging both states should be enough. @darerodz can you open a new issue for that, and another one to also merge contexts?

**Comment by SantosGuillamot** [comment]
> For context here, we also need to merge the state that comes in the new page during client-side navigations.

FYI: I think we will face this issue in [the Movies Demo](https://github.com/c4rl0sbr4v0/wp-movies-demo). There will probably be interactive blocks that aren't on the homepage, and after navigating to a movie, they won't work as expected. I'm thinking of a Tabs block to toggle between images and videos, for example.

**Pull request merged**